drivers/video/video_stm32_dcmi.c

```diff
@@ -0,0 +1,522 @@
+		!device_is_ready(data->dma.dma_dev)) {
+		LOG_ERR("%s DMA device not ready", data->dma.dma_dev->name);
+		return -ENODEV;
+	}
+
+	/*
+	 * DMA configuration
+	 * Due to use of QSPI HAL API in current driver,
+	 * both HAL and Zephyr DMA drivers should be configured.
+	 * The required configuration for Zephyr DMA driver should only provide
+	 * the minimum information to inform the DMA slot will be in used and
+	 * how to route callbacks.
+	 */
+	struct dma_config dma_cfg = data->dma.cfg;
+	static DMA_HandleTypeDef hdma;
+
+	/* Proceed to the minimum Zephyr DMA driver init */
+	dma_cfg.user_data = &hdma;
+	/* HACK: This field is used to inform driver that it is overridden */
+	dma_cfg.linked_channel = STM32_DMA_HAL_OVERRIDE;
+	/* Because of the STREAM OFFSET, the DMA channel given here is from 1 - 8 */
+	ret = dma_config(data->dma.dma_dev,
+			 (data->dma.channel + STM32_DMA_STREAM_OFFSET), &dma_cfg);
+	if (ret != 0) {
+		LOG_ERR("Failed to configure DMA channel %d",
+			data->dma.channel + STM32_DMA_STREAM_OFFSET);
+		return ret;
+	}
+
+	/*** Configure the DMA ***/
+	/* Set the parameters to be configured */
+	hdma.Init.Request		= DMA_REQUEST_DCMI;
+	hdma.Init.Direction		= DMA_PERIPH_TO_MEMORY;
+	hdma.Init.PeriphInc		= DMA_PINC_DISABLE;
+	hdma.Init.MemInc		= DMA_MINC_ENABLE;
+	hdma.Init.PeriphDataAlignment	= DMA_PDATAALIGN_WORD;
+	hdma.Init.MemDataAlignment	= DMA_MDATAALIGN_WORD;
+	hdma.Init.Mode			= DMA_CIRCULAR;
+	hdma.Init.Priority		= DMA_PRIORITY_HIGH;
+	hdma.Init.FIFOMode		= DMA_FIFOMODE_DISABLE;
+
+	hdma.Instance = __LL_DMA_GET_STREAM_INSTANCE(data->dma.reg,
+						data->dma.channel);
+
+	/* Initialize DMA HAL */
+	__HAL_LINKDMA(&data->hdcmi, DMA_Handle, hdma);
+
+	if (HAL_DMA_Init(&hdma) != HAL_OK) {
+		LOG_ERR("DCMI DMA Init failed");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int stm32_dcmi_enable_clock(const struct device *dev)
+{
+	const struct video_stm32_dcmi_config *config = dev->config;
+	const struct device *dcmi_clock = DEVICE_DT_GET(STM32_CLOCK_CONTROL_NODE);
+	int err;
+
+	if (!device_is_ready(dcmi_clock)) {
+		LOG_ERR("clock control device not ready");
+		return -ENODEV;
+	}
+
+	/* Turn on DCMI peripheral clock */
+	err = clock_control_on(dcmi_clock, (clock_control_subsys_t *) &config->pclken);
+	if (err < 0) {
+		LOG_ERR("Failed to enable DCMI clock");
+		LOG_ERR("Error code %d: %s", err, strerror(-err));
+		return err;
+	}
+
+	return 0;
+}
+
+static int video_stm32_dcmi_set_fmt(const struct device *dev,
+				  enum video_endpoint_id ep,
+				  struct video_format *fmt)
+{
+	const struct video_stm32_dcmi_config *config = dev->config;
+	struct video_stm32_dcmi_data *data = dev->data;
+	unsigned int bpp = video_pix_fmt_bpp(fmt->pixelformat);
+
+	if (!bpp || ep != VIDEO_EP_OUT) {
+		return -EINVAL;
+	}
+
+	data->pixel_format = fmt->pixelformat;
+	data->pitch = fmt->pitch;
+	data->height = fmt->height;
+	data->width = fmt->width;
+
+	if (config->sensor_dev && video_set_format(config->sensor_dev, ep, fmt)) {
```

Comment:
`config->sensor_dev` is defined at build time. Rather than testing in most functions, why not testing at build or init ?

Reply:
It's a good point! The `sensor` property has been set to `required: true` in the binding file. Is this enough to guarantee that `config->sensor_dev` will never be `NULL`?

Reply:
Additionally, as suggested, I can add the check below to the `video_stm32_dcmi_init` function and remove the other verification.

```
static int video_stm32_dcmi_init(const struct device *dev)
{
	const struct video_stm32_dcmi_config *config = dev->config;
	struct video_stm32_dcmi_data *data = dev->data;
	int err;

	if (config->sensor_dev == NULL) {
		LOG_ERR("Sensor device cannot be NULL");
		return -ENODEV;
	}
...
```

Please let me know what you think.

Reply:
> The `sensor` property has been set to `required: true` in the binding file. Is this enough to guarantee that `config->sensor_dev` will never be `NULL`?

Indeed. If not provided you'll get an error at dts parsing, so I think we can even spare the check in init function.

Reply:
Removed the check as suggested. Thanks!